### PR TITLE
Maciej Królikowski - UI Design Patterns Challenge

### PR DESCRIPTION
**Pull request by mkkroliks**

### CODE REVIEW

File: MVP Module/Source/Feed Presentation/FeedPresenter.swift [hunk 1]
```diff
@@ -13,17 +13,24 @@ protocol FeedView {
 	func display(_ viewModel: FeedViewModel)
 }
 
+protocol FeedErrorView {
+	func display(_ viewModel: FeedErrorViewModel?)
```
Comment (caiozullo) [inline MVP Module/Source/Feed Presentation/FeedPresenter.swift:17]: Instead of an optional `FeedErrorViewModel?` it'd be best to define the `message: String?` inside the view model as optional. Because the concept of view model in this context is not optional - we always need a view model. But it may contain an error or not.

File: MVP Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 2]
```diff
@@ -8,7 +8,9 @@ protocol FeedViewControllerDelegate {
 	func didRequestFeedRefresh()
 }
 
-public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching, FeedLoadingView {
+public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching, FeedLoadingView, FeedErrorView {
+	@IBOutlet weak var errorView: ErrorView!
```
Comment (caiozullo) [inline MVP Module/Source/Feed UI/Controllers/FeedViewController.swift:12]: This property doesn't need to be `weak`. Also, the setter can be private to prevent clients from changing the instance:

```suggestion
	@IBOutlet private(set) var errorView: ErrorView!
```

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 3]
```diff
@@ -5,6 +5,8 @@
 import UIKit
 
 public final class FeedViewController: UITableViewController, UITableViewDataSourcePrefetching {
+	@IBOutlet weak var errorView: ErrorView!
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:8]: This property doesn't need to be `weak`. Also, the setter can be private to prevent clients from changing the instance:

```suggestion
	@IBOutlet private(set) var errorView: ErrorView!
```

File: MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 4]
```diff
@@ -280,6 +280,19 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+
+		loader.completeFeedLoadingWithError()
+		XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))
+
+		sut.simulateUserInitiatedFeedReload()
+		XCTAssertEqual(sut.errorMessage, nil)
```
Comment (caiozullo) [inline MVP Module/Tests/Feed UI/FeedUIIntegrationTests.swift:293]: Missing test/assertion to check that the error message is nil after tapping on the error view.

File: MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift [hunk 5]
```diff
@@ -280,6 +280,32 @@ final class FeedUIIntegrationTests: XCTestCase {
...
+	func test_errorView_dismissesErrorMessageOnTap() {
+		let (sut, loader) = makeSUT()
+
+		sut.loadViewIfNeeded()
+		XCTAssertEqual(sut.errorMessage, nil)
+
+		loader.completeFeedLoadingWithError()
+		XCTAssertEqual(sut.errorMessage, localized("FEED_VIEW_CONNECTION_ERROR"))
+
+		sut.errorView?.button.simulate(event: .touchUpInside)
```
Comment (caiozullo) [inline MVVM Module/Tests/Feed UI/FeedUIIntegrationTests.swift:305]: The integration test shouldn't be coupled with the view hierarchy such as accessing buttons directly. Instead, create a helper method such as `sut.simulateTapOnError()`

```suggestion
		sut.simulateTapOnError()
```

This way, you can change the view hierarchy without breaking the tests.

File: MVVM Module/Source/Feed Presentation/FeedViewModel.swift [hunk 6]
```diff
@@ -20,12 +20,16 @@ final class FeedViewModel {
 
 	var onLoadingStateChange: Observer<Bool>?
 	var onFeedLoad: Observer<[FeedImage]>?
+	var onFeedLoadWithErrorMessage: Observer<String>?
```
Comment (caiozullo) [inline MVVM Module/Source/Feed Presentation/FeedViewModel.swift:23]: `onFeedLoadWithErrorMessage: Observer<String>?` could be `onErrorStateChange: Observer<String?>?`, so the ViewModel decides when to show/hide errors (instead of the view controller). It's easier to manage presentation state transitions when all changes are centralized.

For example, in the `loadFeed()` method, the ViewModel can call  `onErrorStateChange(.none)` to hide the error before loading the feed.

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift [hunk 7]
```diff
@@ -27,11 +29,15 @@ public final class FeedViewController: UITableViewController, UITableViewDataSou
 		title = viewModel?.title
 		viewModel?.onLoadingStateChange = { [weak self] isLoading in
 			if isLoading {
+				self?.errorView.hideMessage()
```
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:32]: The ViewModel should decide when to hide/show errors. See my other comment.

**Review by caiozullo — COMMENTED** [review]
Hi @mkkroliks, you almost passed the challenge!

Please, check my review comments and let me know once you're done so I can review it again. 

Thank you 🙂

**Comment by mkkroliks** [comment]
Hi @caiozullo,

I applied all your comments.

Thanks for CR ;)

File: MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift (hunk 3 above)
Comment (caiozullo) [inline MVVM Module/Source/Feed UI/Controllers/FeedViewController.swift:8]: This property doesn't need to be `weak`. Also, the setter can be private to prevent clients from changing the instance:

```suggestion
	@IBOutlet private(set) var errorView: ErrorView!
```

**Review by caiozullo — COMMENTED** [review]
Hi @mkkroliks,

Congratulations, you've passed the challenge 🥇

- All tests are passing ✅
- Your code is clean and handles edge cases ✅
- The UI and animations look and behave as expected ✅
- Light and Dark mode support ✅
- Added an extra language ✅

I left only one review comment. Feel free to push changes.

Great job!
